workspace/TS100/Middlewares/Third_Party/FreeRTOS/Source/portable/GCC/ARM_CM3/port.c

```diff
@@ -483,7 +473,7 @@ void xPortSysTickHandler( void )
 
 		/* Enter a critical section but don't use the taskENTER_CRITICAL()
 		method as that will mask interrupts that should exit sleep mode. */
-		__asm volatile( "cpsid i" );
+		__asm volatile( "cpsid i" ::: "memory" );

```

Comment:
Could you give me a 1 liner on what this change achieves?

Reply:
I copied the port files verbatim from FreeRTOS-kernel git. The general idea is that "memory" clobber sections prevent compiler from reordering memory accesses across that line (C standard allows that for non-volatile variables). With -O3 and older port code I was hunting down an odd issue which (after plenty of debugging) boiled down to ``ulTaskNotify()`` leading to critical sections imbalance. Surprisingly, that was almost exactly the scenario explained by David Brown in ``Posted by davidbrown on February 27, 2017`` post on this thread: https://www.freertos.org/FreeRTOS_Support_Forum_Archive/February_2017/freertos_Memory_barriers_in_FreeRTOS_8049c367j.html . I guess after reading that message the FreeRTOS maintainers sprinkled the code with memory barriers liberally.

Reply:
Makes total sense; Thank you for the link.
Mostly was wondering if this was an optimization from you or from upstream so I knew if its something I have to watch out for down the line. If its from upstream then its much easier to not have to worry about it in future freertos upgrades :)